src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,667 @@
+        private void MoveUp((TElement element, TPriority priority) node, int nodeIndex)
+        {
+            // Instead of swapping items all the way to the root, we will perform
+            // a similar optimization as in the insertion sort.
+
+            while (nodeIndex > 0)
+            {
+                int parentIndex = GetParentIndex(nodeIndex);
+                var parent = _nodes[parentIndex];
+
+                if (this.Comparer.Compare(node.priority, parent.priority) < 0)
+                {
+                    PutAt(parent, nodeIndex);
+                    nodeIndex = parentIndex;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            PutAt(node, nodeIndex);
+        }
+
+        /// <summary>
+        /// Moves a node down in the tree to restore heap order.
+        /// </summary>
+        private void MoveDown((TElement element, TPriority priority) node, int nodeIndex)
+        {
+            // The node to move down will not actually be swapped every time.
+            // Rather, values on the affected path will be moved up, thus leaving a free spot
+            // for this value to drop in. Similar optimization as in the insertion sort.
+
+            int i;
+            while ((i = GetFirstChildIndex(nodeIndex)) < _size)
+            {
+                // Check if the current node (pointed by 'nodeIndex') should really be extracted
+                // first, or maybe one of its children should be extracted earlier.
+                var topChild = _nodes[i];
+                int childrenIndexesLimit = (int)Math.Min(i + Arity, _size);
+                int topChildIndex = i;
+
+                while (++i < childrenIndexesLimit)
+                {
+                    var child = _nodes[i];
+                    if (this.Comparer.Compare(child.priority, topChild.priority) < 0)
+                    {
+                        topChild = child;
+                        topChildIndex = i;
+                    }
+                }
+
+                // In case no child needs to be extracted earlier than the current node,
+                // there is nothing more to do - the right spot was found.
+                if (this.Comparer.Compare(node.priority, topChild.priority) <= 0)
+                {
+                    break;
+                }
+
+                // Move the top child up by one node and now investigate the
+                // node that was considered to be the top child (recursive).
+                PutAt(topChild, nodeIndex);
+                nodeIndex = topChildIndex;
+            }
+
+            PutAt(node, nodeIndex);
+        }
+
+        public partial class UnorderedItemsCollection : IReadOnlyCollection<(TElement element, TPriority priority)>, ICollection
+        {
+            private readonly PriorityQueue<TElement, TPriority> _queue;
+
+            public UnorderedItemsCollection(PriorityQueue<TElement, TPriority> queue)
```

Comment:
I think this constructor should be internal. There is no reason to expose its constructor and it was not in the approved shape.